File: 416-partition-equal-subset-sum/partition-equal-subset-sum.cpp
class Solution {
public:
    bool solve(int ind, int target, vector<int>& a, vector<vector<int>>& dp) {
        if (target == 0)
            return true;
        if (ind == 0)
            return a[ind] == target;
        if (dp[ind][target] != -1)
            return dp[ind][target];
        bool nt = solve(ind - 1, target, a, dp);
        bool t = false;
        if (target >= a[ind]) {
            t = solve(ind - 1, target - a[ind], a, dp);
        }
        return dp[ind][target] = t || nt;
    }
    bool canPartition(vector<int>& a) {
        int sum = accumulate(a.begin(), a.end(), 0);
        int n = a.size();
        int target = sum / 2;
        vector<vector<bool>> dp(n, vector<bool>(target + 1, 0));
        if (sum % 2 != 0)
            return false;
        else {
            for (int i = 0; i < n; i++) {
                dp[i][0] = true;
            }
            if (a[0] <= target)
            dp[0][a[0]] = true;
            
            for (int ind = 1; ind < n; ind++) {
                for (int j = 1; j <= target; j++) {
                    bool nt = dp[ind - 1][j];
                    bool t = false;
                    if (j >= a[ind]) {
                        t = dp[ind - 1][j - a[ind]];
                    }
                    dp[ind][j] = t || nt;
                }
            }
            return dp[n - 1][target];
        }
    }
};

File: 416-partition-equal-subset-sum/partition-equal-subset-sum.java
class Solution {
    public int solve(int i,int sum,int n,int[]arr,int[][]dp){
        if(i>=n){
            return 0;
        }
        if(sum==arr[i]){
            return dp[i][sum]=1;
        }
        if(dp[i][sum]!=-1){
            return dp[i][sum];
        }
        int take=0;
        if(arr[i]<=sum){
            take=solve(i+1,sum-arr[i],n,arr,dp);
        }
        int not_take=solve(i+1,sum,n,arr,dp);
        return dp[i][sum]=take | not_take;
    }
    public boolean canPartition(int[] nums) {
        int n=nums.length;
        int sum=0;
        for(int i=0;i<n;i++){
            sum+=nums[i];
        }
        if(sum%2!=0)return false;
        sum = sum/2;
        int [][]dp = new int[n][sum+1];
        for(int i=0;i<n;i++){
            Arrays.fill(dp[i],-1);
        }
        if(solve(0,sum,n,nums,dp)==1){
            return true;
        }else{
            return false;
        }
    }
}